File: aural/whypit-extractor.js
/**
 * Whyp.it Extractor
 *
 * Pure audio extractor - downloads directly to target path provided by caller.
 * Returns platform-agnostic schema.
 */

const { chromium } = require("playwright");
const fs = require("fs").promises;
const path = require("path");
const crypto = require('crypto');

class WhypitExtractor {
  constructor(config = {}) {
    this.platform = 'whypit';
    this.requestDelay = config.requestDelay || 2000;
    this.lastRequestTime = 0;
    this.browser = null;
    this.page = null;
  }

  async setupPlaywright() {
    try {
      console.log("🚀 Starting Playwright browser...");
      this.browser = await chromium.launch({ headless: true });
      this.page = await this.browser.newPage();

      await this.page.setExtraHTTPHeaders({
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36",
      });

      console.log("✅ Playwright browser initialized successfully");
    } catch (error) {
      console.error("❌ Failed to initialize Playwright:", error.message);
      throw error;
    }
  }

  async closeBrowser() {
    if (this.browser) {
      await this.browser.close();
      console.log("🔧 Playwright browser closed");
    }
  }

  /**
   * Extract content from Whyp.it URL
   * @param {string} url - Whyp.it URL to extract
   * @param {object} targetPath - Target paths for output files
   * @param {string} targetPath.dir - Directory to save files
   * @param {string} targetPath.basename - Base filename (without extension)
   */
  async extract(url, targetPath) {
    await this.rateLimit();

    try {
      console.log(`📥 Processing: ${url}`);

      const { dir, basename } = targetPath;
      await fs.mkdir(dir, { recursive: true });

      // Check if already extracted (JSON exists)
      const jsonPath = path.join(dir, `${basename}.json`);
      try {
        await fs.access(jsonPath);
        const cached = JSON.parse(await fs.readFile(jsonPath, 'utf8'));
        console.log(`✅ Using cached extraction for: ${url}`);
        return cached;
      } catch (e) {
        // Not cached, continue with extraction
      }

      await this.page.goto(url, { waitUntil: "networkidle" });

      // Extract metadata from the page
      const pageData = await this.extractPageData();

      // Extract track info from URL
      const urlMatch = url.match(/\/tracks\/(\d+)\/(.+)$/);
      if (!urlMatch) {
        throw new Error("Invalid Whyp.it URL format");
      }

      const trackId = urlMatch[1];
      const titleSlug = urlMatch[2];
      const performer = pageData.performer || "unknown";
      const title = pageData.title || titleSlug;

      // Capture audio URL
      const audioUrl = await this.captureAudioUrl();

      // Download audio
      const audioFilePath = path.join(dir, `${basename}.mp3`);
      await this.downloadFile(audioUrl, audioFilePath);

      // Calculate checksum
      const checksum = await this.calculateChecksum(audioFilePath);

      // Get file stats
      const stats = await fs.stat(audioFilePath);

      // Save HTML backup
      const htmlPath = path.join(dir, `${basename}.html`);
      await this.saveHtmlBackup(htmlPath);

      // Build result in platform-agnostic schema
      const result = {
        audio: {
          sourceUrl: url,
          downloadUrl: audioUrl,
          filePath: audioFilePath,
          format: 'mp3',
          fileSize: stats.size,
          checksum: {
            sha256: checksum
          }
        },
        metadata: {
          title: title,
          author: performer,
          description: pageData.description || '',
          tags: pageData.tags || [],
          duration: null,
          platform: {
            name: 'whypit',
            url: 'https://whyp.it'
          }
        },
        platformData: {
          trackId: trackId,
          titleSlug: titleSlug,
          extractedAt: new Date().toISOString()
        },
        backupFiles: {
          html: htmlPath,
          metadata: jsonPath
        }
      };

      // Save metadata (serves as completion marker)
      await fs.writeFile(jsonPath, JSON.stringify(result, null, 2));

      console.log(`✅ Successfully extracted: ${title} by ${performer}`);
      return result;

    } catch (error) {
      console.error(`❌ Failed to extract ${url}:`, error.message);
      throw error;
    }
  }

  async extractPageData() {
    return await this.page.evaluate(() => {
      // Extract title
      const titleElement = document.querySelector("h1");
      const title = titleElement ? titleElement.textContent.trim() : null;

      // Extract performer
      const userLinkElement = document.querySelector('a[href*="/users/"]');
      let performer = null;
      if (userLinkElement) {
        const performerElement = userLinkElement.querySelector('div:first-child div:first-child');
        if (performerElement) {
          const fullText = performerElement.textContent.trim();
          performer = fullText.split('\n')[0].trim();
        }
      }

      // Extract description and tags
      const metaDescriptionElement = document.querySelector('meta[name="description"]');
      let description = null;
      let tags = [];
      
      if (metaDescriptionElement) {
        description = metaDescriptionElement.getAttribute('content');
        if (description) {
          description = description.trim();
          const tagMatches = description.match(/\[([^\]]+?)\]/g);
          if (tagMatches) {
            tags = tagMatches
              .map(tag => tag.replace(/[\[\]]/g, "").trim())
              .filter(tag => tag.length > 0);
          }
        }
      }

      return {
        title: title,
        performer: performer,
        description: description,
        tags: tags
      };
    });
  }

  async captureAudioUrl() {
    const audioUrls = [];
    
    // Set up network listener
    this.page.on('response', async (response) => {
      const url = response.url();
      if (url.includes('cdn.whyp.it') && url.includes('.mp3') && url.includes('token=')) {
        audioUrls.push(url);
        console.log(`🎵 Found audio URL: ${url}`);
      }
    });

    // Click play button
    console.log("🎯 Looking for play button...");
    
    const playButtonClicked = await this.page.evaluate(() => {
      const buttons = document.querySelectorAll('button');
      for (const button of buttons) {
        const hasPlayIcon = button.innerHTML.includes('play') || 
                           button.querySelector('svg path[d*="8,5.14V19.14L19,12.14L8,5.14Z"]') ||
                           button.querySelector('path[d*="play"]');
        
        if (hasPlayIcon || button.getAttribute('aria-label')?.includes('play')) {
          button.click();
          return true;
        }
      }
      
      const audioPlayerButtons = document.querySelectorAll('button[class*="relative"], button[class*="cursor-pointer"]');
      if (audioPlayerButtons.length > 0) {
        audioPlayerButtons[0].click();
        return true;
      }
      
      return false;
    });

    if (!playButtonClicked) {
      throw new Error("Could not find or click play button");
    }

    console.log("▶️ Play button clicked, waiting for audio URL...");

    // Wait for audio URL
    let waitTime = 0;
    const maxWaitTime = 10000;
    while (audioUrls.length === 0 && waitTime < maxWaitTime) {
      await new Promise(resolve => setTimeout(resolve, 500));
      waitTime += 500;
    }

    if (audioUrls.length === 0) {
      throw new Error("No audio URL found in network requests");
    }

    return audioUrls[0];
  }

  async downloadFile(url, filePath, maxRetries = 5) {
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        console.log(`📥 Download attempt ${attempt}/${maxRetries}...`);
        const response = await fetch(url, {
          headers: {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
          },
        });

        if (!response.ok) {
          throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const buffer = await response.arrayBuffer();
        await fs.writeFile(filePath, Buffer.from(buffer));
        console.log("✅ Audio download completed successfully");
        return true;
      } catch (error) {
        console.log(`❌ Download attempt ${attempt}/${maxRetries} failed: ${error.message}`);
        if (attempt < maxRetries) {
          const waitTime = 5000 * attempt;
          console.log(`⏱️ Waiting ${waitTime / 1000} seconds before retry...`);
          await new Promise((resolve) => setTimeout(resolve, waitTime));
        } else {
          throw error;
        }
      }
    }
  }

  async saveHtmlBackup(htmlPath) {
    const htmlContent = await this.page.content();
    await fs.writeFile(htmlPath, htmlContent);
  }

  async calculateChecksum(filePath) {
    const fileBuffer = await fs.readFile(filePath);
    const hashSum = crypto.createHash('sha256');
    hashSum.update(fileBuffer);
    return hashSum.digest('hex');
  }

  sanitizeFilename(name) {
    return name.replace(/[^A-Za-z0-9 \-_]/g, "");
  }

  async rateLimit() {
    const currentTime = Date.now();
    const timeSinceLastRequest = currentTime - this.lastRequestTime;

    if (timeSinceLastRequest < this.requestDelay) {
      const sleepTime = this.requestDelay - timeSinceLastRequest;
      console.log(`⏱️ Rate limiting: waiting ${sleepTime}ms...`);
      await new Promise((resolve) => setTimeout(resolve, sleepTime));
    }

    this.lastRequestTime = Date.now();
  }
}

module.exports = WhypitExtractor;

File: aural/whypit_extractor.py
#!/usr/bin/env python3
"""
Whyp.it Extractor

Pure audio extractor - downloads directly to target path provided by caller.
Returns platform-agnostic schema.
"""

import argparse
import hashlib
import json
import re
import sys
import time
from datetime import UTC, datetime
from pathlib import Path

import httpx
from dotenv import load_dotenv
from playwright.sync_api import sync_playwright


class WhypitExtractor:
    def __init__(self, config: dict | None = None):
        config = config or {}
        self.platform = "whypit"
        self.request_delay = config.get("request_delay", 2.0)
        self.last_request_time = 0
        self.browser = None
        self.page = None
        self.playwright = None
        self.context = None

    def setup_playwright(self, page=None, context=None):
        """Initialize Playwright browser or use provided page/context."""
        if page is not None:
            self.page = page
            self.context = context
            self._owns_browser = False
            print("🔗 Using shared Playwright browser")
            return

        try:
            print("🚀 Starting Playwright browser...")
            self.playwright = sync_playwright().start()
            self.browser = self.playwright.chromium.launch(headless=True)
            self.context = self.browser.new_context(
                user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36"
            )
            self.page = self.context.new_page()
            self._owns_browser = True
            print("✅ Playwright browser initialized successfully")
        except Exception as error:
            print(f"❌ Failed to initialize Playwright: {error}")
            raise

    def close_browser(self):
        """Close Playwright browser if we own it."""
        if not getattr(self, "_owns_browser", True):
            self.page = None
            self.context = None
            return

        if self.page:
            self.page.close()
            self.page = None
        if self.context:
            self.context.close()
            self.context = None
        if self.browser:
            self.browser.close()
            self.browser = None
        if self.playwright:
            self.playwright.stop()
            self.playwright = None
        print("🔧 Playwright browser closed")

    def extract(self, url: str, target_path: dict) -> dict:
        """
        Extract content from Whyp.it URL.

        Args:
            url: Whyp.it URL to extract
            target_path: Dict with 'dir' and 'basename' keys

        Returns:
            Platform-agnostic metadata dict
        """
        self.ensure_rate_limit()

        try:
            print(f"📥 Processing: {url}")

            target_dir = Path(target_path["dir"])
            basename = target_path["basename"]
            target_dir.mkdir(parents=True, exist_ok=True)

            # Check if already extracted (JSON exists)
            json_path = target_dir / f"{basename}.json"
            if json_path.exists():
                try:
                    cached = json.loads(json_path.read_text(encoding="utf-8"))
                    print(f"✅ Using cached extraction for: {url}")
                    return cached
                except json.JSONDecodeError:
                    print(f"⚠️  Cached JSON unreadable, re-extracting: {url}")

            # Navigate to page (use domcontentloaded instead of networkidle to avoid timeout)
            self.page.goto(url, wait_until="domcontentloaded")
            # Wait a bit for dynamic content
            time.sleep(2)

            # Extract metadata from the page
            page_data = self.extract_page_data()

            # Extract track info from URL (strip query params first)
            clean_url = url.split("?")[0]
            url_match = re.search(r"/tracks/(\d+)/(.+)$", clean_url)
            if not url_match:
                raise ValueError("Invalid Whyp.it URL format")

            track_id = url_match.group(1)
            title_slug = url_match.group(2)
            performer = page_data.get("performer") or "unknown"
            title = page_data.get("title") or title_slug

            # Capture audio URL
            audio_url = self.capture_audio_url()

            # Determine audio format from URL
            clean_audio_url = audio_url.split("?")[0]
            if clean_audio_url.endswith(".flac"):
                audio_format = "flac"
            else:
                audio_format = "mp3"

            # Download audio
            audio_file_path = target_dir / f"{basename}.{audio_format}"
            self.download_file(audio_url, audio_file_path)

            # Calculate checksum
            checksum = self.calculate_checksum(audio_file_path)

            # Get file stats
            stats = audio_file_path.stat()

            # Save HTML backup
            html_path = target_dir / f"{basename}.html"
            self.save_html_backup(html_path)

            # Build result in platform-agnostic schema
            result = {
                "audio": {
                    "sourceUrl": url,
                    "downloadUrl": clean_audio_url,
                    "filePath": str(audio_file_path),
                    "format": audio_format,
                    "fileSize": stats.st_size,
                    "checksum": {"sha256": checksum},
                },
                "metadata": {
                    "title": title,
                    "author": performer,
                    "description": page_data.get("description") or "",
                    "tags": page_data.get("tags") or [],
                    "duration": None,
                    "platform": {"name": "whypit", "url": "https://whyp.it"},
                },
                "platformData": {
                    "trackId": track_id,
                    "titleSlug": title_slug,
                    "extractedAt": datetime.now(UTC)
                    .isoformat()
                    .replace("+00:00", "Z"),
                },
                "backupFiles": {"html": str(html_path), "metadata": str(json_path)},
            }

            # Save metadata (serves as completion marker)
            json_path.write_text(
                json.dumps(result, indent=2, ensure_ascii=False), encoding="utf-8"
            )

            print(f"✅ Successfully extracted: {title} by {performer}")
            return result

        except Exception as error:
            print(f"❌ Failed to extract {url}: {error}")
            raise

    def extract_page_data(self) -> dict:
        """Extract metadata from Whyp.it page."""
        return self.page.evaluate(
            """() => {
            // Extract title
            const titleElement = document.querySelector("h1");
            const title = titleElement ? titleElement.textContent.trim() : null;

            // Extract performer from uploader link
            // The uploader link contains "tracks" text (e.g., "598 tracks")
            let performer = null;
            const userLinks = document.querySelectorAll('a[href*="/users/"]');
            for (const link of userLinks) {
                if (link.textContent.includes('tracks')) {
                    const href = link.getAttribute('href');
                    // Format: /users/4994/lurkydip
                    const match = href.match(/\\/users\\/\\d+\\/([^/?]+)/);
                    if (match) {
                        performer = match[1];
                    }
                    break;
                }
            }

            // Extract description and tags from meta description
            const metaDescriptionElement = document.querySelector('meta[name="description"]');
            let description = null;
            let tags = [];

            if (metaDescriptionElement) {
                description = metaDescriptionElement.getAttribute('content');
                if (description) {
                    description = description.trim();

                    // Extract tags from [tag] format
                    const tagMatches = description.match(/\\[([^\\]]+?)\\]/g);
                    if (tagMatches) {
                        tags = tagMatches
                            .map(tag => tag.replace(/[\\[\\]]/g, "").trim())
                            .filter(tag => tag.length > 0);
                    }
                }
            }

            return {
                title: title,
                performer: performer,
                description: description,
                tags: tags
            };
        }"""
        )

    def capture_audio_url(self) -> str:
        """Capture audio URL by clicking play button and intercepting network requests."""
        print("🎯 Looking for play button...")

        def is_audio_response(r):
            """Check if response is an audio file (mp3 or flac)."""
            if "cdn.whyp.it" not in r.url:
                return False
            url_path = r.url.split("?")[0]
            return url_path.endswith((".mp3", ".flac"))

        # Use expect_response to wait for audio URL (mp3 or flac) while clicking play
        with self.page.expect_response(is_audio_response, timeout=30000) as response_info:
            play_button_clicked = self.page.evaluate(
                """() => {
                const buttons = document.querySelectorAll('button');
                for (const button of buttons) {
                    const hasPlayIcon = button.innerHTML.includes('play') ||
                                       button.querySelector('svg path[d*="8,5.14V19.14L19,12.14L8,5.14Z"]') ||
                                       button.querySelector('path[d*="play"]');

                    if (hasPlayIcon || button.getAttribute('aria-label')?.includes('play')) {
                        button.click();
                        return true;
                    }
                }

                const audioPlayerButtons = document.querySelectorAll('button[class*="relative"], button[class*="cursor-pointer"]');
                if (audioPlayerButtons.length > 0) {
                    audioPlayerButtons[0].click();
                    return true;
                }

                return false;
            }"""
            )

            if not play_button_clicked:
                raise ValueError("Could not find or click play button")

            print("▶️ Play button clicked, waiting for audio URL...")

        response = response_info.value
        audio_url = response.url
        # Strip query parameters for logging (keep full URL for download)
        clean_url = audio_url.split("?")[0]
        print(f"🎵 Found audio URL: {clean_url}")

        return audio_url

    def download_file(self, url: str, file_path: Path, max_retries: int = 5):
        """Download file from URL with retries using streaming."""
        for attempt in range(1, max_retries + 1):
            try:
                print(f"📥 Download attempt {attempt}/{max_retries}...")

                with httpx.stream(
                    "GET",
                    url,
                    headers={
                        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
                    },
                    follow_redirects=True,
                    timeout=60.0,
                ) as response:
                    response.raise_for_status()
                    with file_path.open("wb") as f:
                        for chunk in response.iter_bytes(chunk_size=8192):
                            f.write(chunk)

                print("✅ Audio download completed successfully")
                return True

            except Exception as error:
                print(f"❌ Download attempt {attempt}/{max_retries} failed: {error}")
                if attempt < max_retries:
                    wait_time = 5 * attempt
                    print(f"⏱️ Waiting {wait_time} seconds before retry...")
                    time.sleep(wait_time)
                else:
                    raise

    def save_html_backup(self, html_path: Path):
        """Save HTML content as backup."""
        html_content = self.page.content()
        html_path.write_text(html_content, encoding="utf-8")

    def calculate_checksum(self, file_path: Path) -> str:
        """Calculate SHA256 checksum of file."""
        file_buffer = file_path.read_bytes()
        hash_sum = hashlib.sha256()
        hash_sum.update(file_buffer)
        return hash_sum.hexdigest()

    def sanitize_filename(self, name: str) -> str:
        """Sanitize filename by removing invalid characters."""
        return re.sub(r"[^A-Za-z0-9 \-_]", "", name)

    def ensure_rate_limit(self):
        """Ensure rate limiting between requests."""
        now = time.time()
        time_since_last_request = now - self.last_request_time

        if time_since_last_request < self.request_delay:
            delay = self.request_delay - time_since_last_request
            print(f"⏳ Rate limiting: waiting {delay * 1000:.0f}ms...")
            time.sleep(delay)

        self.last_request_time = time.time()


def main():
    """CLI entry point for testing."""
    parser = argparse.ArgumentParser(
        description="Extract audio content from Whyp.it URLs"
    )
    parser.add_argument("url", help="Whyp.it URL to extract")
    parser.add_argument(
        "--output-dir",
        "-o",
        default="data/whypit",
        help="Output directory (default: data/whypit)",
    )
    parser.add_argument(
        "--basename",
        "-b",
        help="Base filename (without extension). If not provided, extracts from URL.",
    )

    args = parser.parse_args()

    # Load environment variables
    load_dotenv()

    # Extract basename from URL if not provided
    basename = args.basename
    if not basename:
        # Strip query parameters before extracting basename
        clean_url = args.url.split("?")[0]
        url_match = re.search(r"/tracks/\d+/(.+)$", clean_url)
        if url_match:
            basename = url_match.group(1)
        else:
            print("❌ Could not extract basename from URL. Please provide --basename")
            return 1

    target_path = {"dir": args.output_dir, "basename": basename}

    extractor = WhypitExtractor()

    try:
        extractor.setup_playwright()
        result = extractor.extract(args.url, target_path)
        print("\n✅ Extraction complete!")
        print(f"📁 Files saved to: {target_path['dir']}")
        print(f"🎵 Audio: {result['audio']['filePath']}")
        print(f"📄 Metadata: {result['backupFiles']['metadata']}")
        return 0
    except Exception as e:
        print(f"\n❌ Error: {e}")
        return 1
    finally:
        extractor.close_browser()


if __name__ == "__main__":
    sys.exit(main())
